File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(root, mini, maxi):
            # important
            if not root:
                return True
            # important to compare mini and maxi with None
            if mini != None and root.val <= mini:
                return False
            if maxi != None and root.val >= maxi:
                return False
            # impartant to return
            return dfs(root.left, mini, root.val) and dfs(root.right, root.val, maxi)
        # important to return
        return dfs(root, None, None)
    

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean helperIsValidBST(TreeNode root, Integer mini, Integer maxi) {
        if(root == null) return true;

        if (mini != null && root.val <= mini) return false;

        if (maxi != null && root.val >= maxi) return false;

        return helperIsValidBST(root.left, mini, root.val) && helperIsValidBST(root.right, root.val, maxi);

    }
    public boolean isValidBST(TreeNode root) {

        return helperIsValidBST(root, null, null);
        
    }
}